server/src/main/java/org/elasticsearch/cluster/coordination/JoinHelper.java

```diff
@@ -0,0 +1,214 @@
+                // the term under which this state will eventually be published: the current term may be increased after this check due to
+                // some other activity. That the term is correct is, however, checked properly during publication, so it is sufficient to
+                // check it here on a best-effort basis. This is fine because a concurrent change indicates the existence of another leader
+                // in a higher term which will cause this node to stand down.
+
+                final long currentTerm = currentTermSupplier.getAsLong();
+                if (currentState.term() != currentTerm) {
+                    currentState = ClusterState.builder(currentState).term(currentTerm).build();
+                }
+                return super.execute(currentState, joiningTasks);
+            }
+
+        };
+
+        transportService.registerRequestHandler(JOIN_ACTION_NAME, ThreadPool.Names.GENERIC, false, false, JoinRequest::new,
+            (request, channel, task) -> joinHandler.accept(request, new JoinCallback() {
+
+                @Override
+                public void onSuccess() {
+                    try {
+                        channel.sendResponse(TransportResponse.Empty.INSTANCE);
+                    } catch (IOException e) {
+                        onFailure(e);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    try {
+                        channel.sendResponse(e);
+                    } catch (Exception inner) {
+                        inner.addSuppressed(e);
+                        logger.warn("failed to send back failure on join request", inner);
+                    }
+                }
+
+                @Override
+                public String toString() {
+                    return "JoinCallback{request=" + request + "}";
+                }
+            }));
+    }
+
+    public interface JoinCallback {
+        void onSuccess();
+
+        void onFailure(Exception e);
+    }
+
+    static class JoinTaskListener implements ClusterStateTaskListener {
+        private final JoinTaskExecutor.Task task;
+        private final JoinCallback joinCallback;
+
+        JoinTaskListener(JoinTaskExecutor.Task task, JoinCallback joinCallback) {
+            this.task = task;
+            this.joinCallback = joinCallback;
+        }
+
+        @Override
+        public void onFailure(String source, Exception e) {
+            joinCallback.onFailure(e);
+        }
+
+        @Override
+        public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+            joinCallback.onSuccess();
+        }
+
+        @Override
+        public String toString() {
+            return "JoinTaskListener{task=" + task + "}";
+        }
+    }
+
+    interface JoinAccumulator {
```

Comment:
This definitely feels like overkill now the `JoinHelper` is mode-aware and its mode is in sync with the coordinator.